src/Common/src/TypeSystem/IL/ILImporter.cs

```diff
@@ -876,17 +896,25 @@ private void ImportBasicBlock(BasicBlock basicBlock)
                         ImportReadOnlyPrefix();
                         continue;
                     default:
-                        throw new BadImageFormatException("Invalid opcode");
+                        ReportInvalidInstruction(opCode);
+                        EndImportingInstruction();
+                        return;
                 }
 
                 // Check if control falls through the end of method.
-                if (_currentOffset >= _basicBlocks.Length)
-                    throw new TypeSystemException.InvalidProgramException();
-
-                BasicBlock nextBasicBlock = _basicBlocks[_currentOffset];
-                if (nextBasicBlock != null)
+                if (_currentOffset < _basicBlocks.Length)
+                {
+                    BasicBlock nextBasicBlock = _basicBlocks[_currentOffset];
+                    if (nextBasicBlock != null)
+                    {
+                        ImportFallthrough(nextBasicBlock);
+                        return;
```

Comment:
I think it would be best to call `EndImportingInstruction` before the fall-through logic - before the `Check if control falls through the end of method.` comment.